File: Andromeda.AvaloniaApp/ViewModels/ViewModelBase.cs
using Avalonia.Controls;
using ReactiveUI;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Andromeda.AvaloniaApp.Helpers;
using Andromeda.AvaloniaApp.ViewModels.Windows;
using Andromeda.AvaloniaApp.Windows;

using static Andromeda.Core.FSharp.AppData;
using static Andromeda.Core.FSharp.Installed;

namespace Andromeda.AvaloniaApp.ViewModels {
    public class ViewModelBase : ReactiveObject {
        public AppDataWrapper AppDataWrapper { get; private set; }
        protected AppData AppData {
            get => this.AppDataWrapper.AppData;
        }

        private readonly IList<SubViewModelBase> children = new List<SubViewModelBase>();
        public IList<SubViewModelBase> Children { get => this.children; }

        public Control Control { get; }

        public ViewModelBase(Control control) : this(control, null) {
            // Nothing to do here
        }

        public ViewModelBase(Control control, AppDataWrapper appDataWrapper) {
            this.Control = control;

            if (appDataWrapper != null) {
                this.AppDataWrapper = appDataWrapper;
            }
            else {
                this.AppDataWrapper = new AppDataWrapper();
                this.SetAppData(loadAppData());
                this.SetAppData(searchInstalled(this.AppData));
                if (this.AppData.authentication.IsNoAuth) {
                    // Authenticate
                    var window = new AuthenticationWindow();
                    window.DataContext = new AuthenticationWindowViewModel(window, this.AppDataWrapper);
                    window.ShowDialog(this.GetParentWindow());
                }
            }
        }

        public virtual Window GetParentWindow() {
            return this.Control is Window ? (Window)this.Control : null;
        }

        protected virtual void SetAppData(AppData appData) {
            this.AppDataWrapper.AppData = appData;
            saveAppData(this.AppData);
        }

        public IList<T> GetChildrenOfType<T>() {
            return this.children
                .Where(child => typeof(T).IsInstanceOfType(child))
                .Cast<T>()
                .ToList();
        }
    }
}

File: src/Andromeda/AvaloniaApp.FSharp/ViewModels/ViewModelBase.fs
namespace Andromeda.AvaloniaApp.FSharp.ViewModels

open Andromeda.Core.FSharp
open Avalonia.Controls
open ReactiveUI

open Andromeda.AvaloniaApp.FSharp.Helpers

[<AbstractClass>]
type ViewModelBase(appDataWrapper: AppDataWrapper) as this =
    inherit ReactiveObject()

    member val Children: SubViewModelBase list = [] with get, set
    member val AppDataWrapper: AppDataWrapper = appDataWrapper with get, set
    member this.AppData
        with get () = this.AppDataWrapper.AppData

    abstract member GetParentWindow: unit -> Window
    abstract member GetRootViewModel: unit -> ParentViewModelBase

    member __.Init () =
        for child in this.Children do
            child.Init()

    member this.SetAppData (appData): unit =
        this.AppDataWrapper.AppData <- appData
        // TODO: error handling
        AppDataPersistence.save appData |> ignore

    member this.GetChildrenOfType<'T when 'T :> SubViewModelBase> () =
        this.Children
        |> List.fold (fun list child ->
            match child with
            | :? 'T as child -> child::list
            | _ -> list
        ) []

and [<AbstractClass>]
    ParentViewModelBase(window: Window, appDataWrapper) as this =
    inherit ViewModelBase(appDataWrapper)

    member val Control = window

    abstract member AddNotification: string -> unit

    override __.GetParentWindow () = window
    override __.GetRootViewModel () = this

// Think about making this generic and add interfaces for ViewModelBase and SubViewModelBase
and [<AbstractClass>]
    SubViewModelBase(control: Control, parent: ViewModelBase) as this =
    inherit ViewModelBase(parent.AppDataWrapper)

    do parent.Children <- this::parent.Children

    member val Control = control
    member val Parent: ViewModelBase = parent

    override this.GetParentWindow () =
        match this.Control with
        | :? Window as window -> window
        | _ -> parent.GetParentWindow ()

    override __.GetRootViewModel () = this.Parent.GetRootViewModel ()
